SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialLoadBalanceEnforcer.cpp

```diff
@@ -1,271 +1,214 @@
-	{
-		UE_LOG(LogSpatialLoadBalanceEnforcer, Log, TEXT("Entity %lld removed. Can no longer enforce the previous request for this entity."),
-			   Op.entity_id);
-		AclWriteAuthAssignmentRequests.Remove(Op.entity_id);
-	}
-}
-
-void SpatialLoadBalanceEnforcer::OnAclAuthorityChanged(const Worker_AuthorityChangeOp& AuthOp)
-{
-	// This class should only be informed of ACL authority changes.
-	check(AuthOp.component_id == SpatialConstants::ENTITY_ACL_COMPONENT_ID);
+		bool bRefresh = false;
+		switch (Delta.Type)
+		{
+		case EntityDelta::UPDATE:
+		{
+			for (const ComponentChange& Change : Delta.ComponentUpdates)
+			{
+				bRefresh |= ApplyComponentUpdate(Delta.EntityId, Change.ComponentId, Change.Update);
+			}
+			for (const ComponentChange& Change : Delta.ComponentsRefreshed)
+			{
+				bRefresh |= ApplyComponentRefresh(Delta.EntityId, Change.ComponentId, Change.CompleteUpdate.Data);
+			}
+			break;
+		}
+		case EntityDelta::ADD:
+			PopulateDataStore(Delta.EntityId);
+			bRefresh = true;
+			break;
+		case EntityDelta::REMOVE:
+			DataStore.Remove(Delta.EntityId);
+			break;
+		case EntityDelta::TEMPORARILY_REMOVED:
+			DataStore.Remove(Delta.EntityId);
+			PopulateDataStore(Delta.EntityId);
+			bRefresh = true;
+			break;
+		default:
+			break;
+		}
 
-	if (AuthOp.authority != WORKER_AUTHORITY_AUTHORITATIVE)
-	{
-		if (AclAssignmentRequestIsQueued(AuthOp.entity_id))
+		if (bRefresh)
 		{
-			UE_LOG(LogSpatialLoadBalanceEnforcer, Log,
-				   TEXT("ACL authority lost for entity %lld. Can no longer enforce the previous request for this entity."),
-				   AuthOp.entity_id);
-			AclWriteAuthAssignmentRequests.Remove(AuthOp.entity_id);
+			RefreshAcl(Delta.EntityId);
 		}
-		return;
 	}
-
-	MaybeQueueAclAssignmentRequest(AuthOp.entity_id);
 }
 
-// MaybeQueueAclAssignmentRequest is called from three places.
-// 1) AuthorityIntent change - Intent is not authoritative on this worker - ACL is authoritative on this worker.
-//    (another worker changed the intent, but this worker is responsible for the ACL, so update it.)
-// 2) ACL change - Intent may be anything - ACL just became authoritative on this worker.
-//    (this worker just became responsible, so check to make sure intent and ACL agree.)
-// 3) AuthorityIntent change - Intent is authoritative on this worker but no longer assigned to this worker - ACL is authoritative on this
-// worker.
-//    (this worker had responsibility for both and is giving up authority.)
-// Queuing an ACL assignment request may not occur if the assignment is the same as before, or if the request is already queued,
-// or if we don't meet the predicate required to enforce the assignment.
-void SpatialLoadBalanceEnforcer::MaybeQueueAclAssignmentRequest(const Worker_EntityId EntityId)
+void SpatialLoadBalanceEnforcer::ShortCircuitMaybeRefreshAcl(const Worker_EntityId EntityId)
 {
-	if (!CanEnforce(EntityId))
+	const EntityViewElement& Element = SubView->GetView()[EntityId];
```

Comment:
Seems _pretty_ risky assuming the entity is valid, prefer Find or FindRef

Reply:
This entity should exist in the view if this function is called. Can add a check before if that makes it clearer? I'd say it's incorrect to call this on an entity that doesn't exist.

Reply:
Could replace it with FindChecked if you want to be explicit (which is just what [] calls internally). But I agree that it would make no sense for this entity to not exist.

Reply:
Conceivably could happen if we changed things to start calling this stuff on actors that have been created for which entities don't yet exist, but I don't think we should do that as then it puts the responsibility on systems like this to detect when entities do exist and queue stuff up till then.